### ISSUE TITLE
Ensure transactions never leave the user without a '.flox' directory

### ISSUE THREAD

**Issue by zmitchell** [flox/flox#231]
## Observations
- #230 implements the first iteration of our next (feature flagged) `flox install` command.
- Package installs are intended to be atomic and use transactions to ensure that an installation completely succeeds, or completely fails without making changes to the project.
- The transaction mechanism (for a successful install) looks like this:
    - `<proj>/.flox` is renamed to `<backup_dir>/.flox` as a backup
    - At this point there is no longer a `.flox` in the project directory
    - `<replacement>/.flox` (already verified to build) is renamed to `<proj>/.flox`
    - At this point the project has a `.flox` again

## Problem
- There is a very brief moment in which the project no longer has a `.flox` directory.
- If something happens during that period, the user is left without a `.flox` directory.
- Without a `.flox` directory the user will find that many of the `flox` commands no longer work.

How serious is this problem?
- The likelihood of this problem happening is low (requires something like a power outage on a plugged in machine, a filesystem failure, etc).
- The severity of the problem is also low _if the user is using source control_ since they can simply restore the directory from the index.
    - If the user isn't using source control, they've lost all the effort they spent creating their environment, setting up remotes, pinning package versions, etc.
    - They're likely to be pretty unhappy in that case, even if they could have prevented their own problem with source control.

## Approaches
A more robust transaction mechanism would look like this:
- Copy (instead of `mv`) `<proj>/.flox`  to `<backup>/.flox`
    - Copying isn't atomic like renaming is, so additional steps are required to ensure that the copy was successful. That part is the next step
- Verify that `<backup>/.flox` contains the same contents as `<proj>/.flox`
    - This is supposed to be a backup, so we should ensure that it's a faithful reproduction of the original data
    - You can accomplish this part using `find`, `xargs`, and `md5` (or similar) to run a checksum on every file in the directory, then compute the checksum of the checksums. A quick search online will show you some (long) one-liners that do this in Bash.
    - The `.flox` directory doesn't contain many files, and the files it does contain are relatively small, so I don't foresee this causing a large performance penalty.
- Rename the replacement to `<proj>/.flox`

## Acceptance criteria
- Any operation that performs a transaction (install, uninstall, edit, etc) uses a technique that never leaves a project without a `.flox` directory.

**Comment by mkenigs** [comment]
- `.flox` always exists; it's just `name` that gets renamed to `name.tmp`. So I think you mean `.flox/name.tmp` rather than `<backup>/.flox` throughout
- I don't believe what you're describing is possible, because the new directory for the rename syscall must be empty:
  ```
  oldpath can specify a directory.  In this case, newpath must either not exist, or it must specify an empty directory.
  ```
  https://man7.org/linux/man-pages/man2/rename.2.html
- an alternative approach would be to check for `name.tmp` when we're creating a PathEnvironment object

**Comment by zmitchell** [comment]
> .flox always exists; it's just name that gets renamed to name.tmp. So I think you mean .flox/name.tmp rather than <backup>/.flox throughout

Yeah, you're absolutely right, I keep getting those two mixed up. That certainly makes things feel less catastrophic, especially given that the backup is stored alongside the original directory rather than in `/tmp` somewhere where the user wouldn't be able to easily find it.

When we get things implemented and more stable we should decide on some terminology to refer to each of these pieces more clearly e.g. `.flox` vs. `.flox/name`.

> I don't believe what you're describing is possible, because the new directory for the rename syscall must be empty

I don't see what this has to do with the bug at all. When you rename the old directory it ceases to exist under the old name. That's the issue I was pointing out. It's as if `.flox/name` no longer exists. If something goes wrong before you can restore the backup or before you rename the replacement there's no longer a `.flox/name` directory.

**Comment by mkenigs** [comment]
> I don't see what this has to do with the bug at all. When you rename the old directory it ceases to exist under the old name. That's the issue I was pointing out. It's as if `.flox/name` no longer exists. If something goes wrong before you can restore the backup or before you rename the replacement there's no longer a `.flox/name` directory.

I was responding to your suggested approach. The "bug" is a limitation of Linux filesystems

**Comment by zmitchell** [comment]
> I was responding to your suggested approach.

Ah, I see. In that case we could just use `rsync` to copy the backup, it has a flag to force comparing the checksum of the copied files. I would have to look into how (or whether) it handles symlinks.

> The "bug" is a limitation of Linux filesystems

I don't see this as a filesystem bug. If you move a directory and don't leave anything in its place, the filesystem is doing what you asked it to do.

**Comment by zmitchell** [comment]
Discussed with @mkenigs offiline.

Highlights:
- I was mixed up on the fact that `.flox` contains `.flox/<env name>` directories, and the contents of those project directories are what get replaced during a transaction, not the entire `.flox` directory.
- Given that it's not the entire `.flox` directory that's replaced, and given that the backup is stored alongside the original contents of the project directory, a failure is much less dire.
- If a transaction fails, you can't use `mv` to replace the existing `.flox/<name>` directory because the destination must be empty or not exist.
- Using `rsync` to copy the backup to the original location isn't atomic, and could also have a partial failure, even though the probability of the original failure followed by a subsequent failure is very, very low.
- In this case you no longer know whether the transaction has succeeded.
- I proposed a lockfile, the presence of which would indicate that a transaction was either in progress or had failed to complete.
- @mkenigs pointed out that the presence of `.flox/<name>.tmp` serves the same purpose.
- At this point we agreed that it's fine to momentarily leave the `.flox/<name>` directory missing, and use the presence of `.flox/<name>.tmp` as an indicator as to whether the transaction failed.

**Comment by zmitchell** [comment]
Maybe `.bak` would be a better extension than `.tmp` since `.bak` is commonly used on *nix systems. Also, a user might see `.tmp` and think it's a temporary file that they're free to delete, whereas `.bak` clearly conveys that it's a backup and they might lose something if they delete it.

**Comment by mkenigs** [comment]
> Maybe `.bak` would be a better extension than `.tmp` since `.bak` is commonly used on *nix systems. Also, a user might see `.tmp` and think it's a temporary file that they're free to delete, whereas `.bak` clearly conveys that it's a backup and they might lose something if they delete it.

👍

**Comment by limeytexan** [comment]
Matthew the details in this issue are not quite as we discussed, e.g. it was the `.flox/<name>` directory that was being renamed and moved with a transaction, not the top level `.flox`. We also discussed that the ".bak" directory could be automatically moved back into place at the beginning of any flox invocation that uses that [named] environment (with a warning message).

Let's schedule a time to talk through this in detail.

**Comment by zmitchell** [comment]
Outdated
